src/combinatorics/simple_graph/inc_matrix.lean

```diff
@@ -0,0 +1,415 @@
+              set.not_not_mem, forall_true_left, not_false_iff, zero_ne_one]
+
+-- ### One - zero properties
+
+lemma inc_matrix_not_zero {i : V} {e : G.edge_set} [char_zero R] :
+  ¬ G.inc_matrix R i e = 0 ↔ G.inc_matrix R i e = 1 :=
+by simp only [inc_matrix_zero, inc_matrix_one, set.not_not_mem]
+
+lemma inc_matrix_not_one {i : V} {e : G.edge_set} [char_zero R] :
+  ¬ G.inc_matrix R i e = 1 ↔ G.inc_matrix R i e = 0 :=
+by simp only [inc_matrix_zero, inc_matrix_one]
+
+lemma inc_matrix_zero_or_one {i : V} {e : G.edge_set} [char_zero R] :
+  G.inc_matrix R i e = 0 ∨ G.inc_matrix R i e = 1 :=
+by { rw [inc_matrix_zero, inc_matrix_one], exact (em (e.val ∈ G.incidence_set i)).symm }
+
+lemma inc_matrix_elements_mul_one {i j : V} {e : G.edge_set} [char_zero R] :
+  G.inc_matrix R i e * G.inc_matrix R j e = 1 ↔ G.inc_matrix R i e = 1 ∧ G.inc_matrix R j e = 1 :=
+begin
+  cases G.inc_matrix_zero_or_one R with H₀ H₁,
+  { rw H₀, simp only [if_t_t, mul_boole, inc_matrix_apply, zero_ne_one, false_and] },
+  { rw H₁, simp only [true_and, mul_boole, inc_matrix_apply, eq_self_iff_true] }
+end
+
+-- 1. ∑ e : E, M i e * M j e = 1, where i and j are adjacent.
+theorem adj_sum_of_mul_inc_one {i j : V} (H_adj : G.adj i j) :
+  ∑ (e : G.edge_set), G.inc_matrix R i e * G.inc_matrix R j e = (1 : R) :=
+begin
+  simp only [inc_matrix_apply, ite_mul_ite_zero_right, sum_boole,
+             G.mem_incidence_sets_iff_eq H_adj, ←subtype.val_eq_coe, mul_one],
+  rw adj_iff_exists_edge_val at H_adj,
+  rcases H_adj with ⟨e, H_e⟩,
+  simp only [←H_e, subtype.ext_iff_val.symm],
+  have H : filter (λ (x : G.edge_set), x = e) univ = {e},
+  { ext, simp only [true_and, mem_filter, mem_univ, mem_singleton] },
+  simp only [H, filter_congr_decidable, nat.cast_one, card_singleton]
+end
+
+-- 2. M i e * M j e = 0, where i, j distinct non-adjacent vertices, e an edge.
+theorem inc_matrix_mul_non_adj {i j : V} {e : G.edge_set} (Hne : i ≠ j) (H_non_adj : ¬ G.adj i j)
+[char_zero R] :
+  G.inc_matrix R i e * G.inc_matrix R j e = 0 :=
+begin
+  by_cases H₁ : G.inc_matrix R i e = 0,
+  { rw [H₁, zero_mul] },
+  { rw [inc_matrix_not_zero, inc_matrix_one] at H₁,
+    by_cases H₂ : G.inc_matrix R j e = 0,
+    { rw [H₂, mul_zero] },
+    { rw [inc_matrix_not_zero, inc_matrix_one] at H₂,
+      exfalso,
+      apply H_non_adj,
+      rw [←mem_edge_set, ←G.edge_in_two_incidence_sets Hne ⟨H₁, H₂⟩],
+      exact G.edge_val_in_set } }
+end
+
+-- 3. (M i e) ^ 2 = M i e; with i a vertex, e an edge.
+theorem inc_matrix_element_power_id {i : V} {e : G.edge_set} :
+  (G.inc_matrix R i e) * (G.inc_matrix R i e) = G.inc_matrix R i e :=
+by simp [inc_matrix_apply]
+
+-- 4. degree(i) = ∑ e : E, M i e; where i is a vertex.
+theorem degree_equals_sum_of_incidence_row {i : V} [char_zero R] :
+  (G.degree i : R) = ∑ (e : G.edge_set), G.inc_matrix R i e :=
+begin
+  simp only [inc_matrix_apply, sum_boole],
+  rw [←card_incidence_set_eq_degree, nat.cast_inj],
+  refine finset.card_congr _ _ _ _,
```

Comment:
this is super messy; I wish there was some way to use `fintype.filter_mem_eq_inter`, but it's really hard with `simple_graph` making such heavy use of subtypes. this filter should, imo, not be `(λ x : subtype, _) univ` but `(λ x : sym2 v, _) incidence_set`. anyways, that's just a rant, don't mind me

Reply:
i still think there's some way to make it work, but I'm really not sure

Reply:
I could not manage to make further progress here unfortunately

Reply:
Using the `sym2  V` suggestion:
```lean
@[simp]
lemma card_incidence_finset_eq_degree [decidable_eq V] (v : V) :
  (G.incidence_finset v).card = G.degree v :=
begin
  convert G.card_incidence_set_eq_degree v using 1,
  apply set.to_finset_card,
end

theorem degree_equals_sum_of_incidence_row {i : V} [char_zero R] :
  (G.degree i : R) = ∑ e in G.edge_finset, G.inc_matrix R i e :=
begin
  simp only [inc_matrix_apply, sum_boole, nat.cast_inj],
  rw ←card_incidence_finset_eq_degree,
  apply congr_arg finset.card,
  ext e,
  simp only [mem_filter, iff_and_self, mem_incidence_finset, mem_edge_finset],
  apply G.incidence_set_subset,
end
```